File: 2021/11/09/MinAbsDifference.java
/**
 * https://leetcode.com/problems/minimum-absolute-difference/
 * @author Andrew Jarombek
 * @since 11/9/2021
 */

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class MinAbsDifference {
    public List<List<Integer>> minimumAbsDifference(int[] arr) {
        Arrays.sort(arr);

        int minDiff = Integer.MAX_VALUE;
        int prev = arr[0];
        List<List<Integer>> result = new ArrayList<>();

        for (int i = 1; i < arr.length; i++) {
            int current = arr[i];

            if (current - prev < minDiff) {
                minDiff = current - prev;
                result = new ArrayList<>();
                this.addPair(result, prev, current);
            } else if (current - prev == minDiff) {
                this.addPair(result, prev, current);
            }

            prev = current;
        }

        return result;
    }

    private void addPair(List<List<Integer>> result, int prev, int current) {
        List<Integer> pair = new ArrayList<>();
        pair.add(prev);
        pair.add(current);
        result.add(pair);
    }
}

File: 2022/10/09/minAbsDifference.ts
/**
 * https://leetcode.com/problems/minimum-absolute-difference/
 * @author Andrew Jarombek
 * @since 10/9/2022
 */

function minimumAbsDifference(arr: number[]): number[][] {
    arr.sort((a, b) => a - b);

    let min = Infinity;
    let result = [];
    let prev = arr[0];

    for (let val of arr.slice(1)) {
        const diff = val - prev;

        if (diff < min) {
            result = [[prev, val]];
            min = diff;
        } else if (diff == min) {
            result.push([prev, val]);
        }

        prev = val;
    }

    return result;
}